Lots of defect fixes

tambry opened the pull request:
Includes a few memory corruption bugs fixed, added initialization for a couple variables, fixed some flow control problems and added error checks and error logging to a lot of places.

HDD functions now also return a negative version, if they fail seeking. Didn't add any checks for when they're actually used, but if we'd want to be more or less correct, we should check for errors from those functions in the future.

vlj commented on rpcs3/Emu/HDD/HDD.cpp:
```diff
@@ -19,7 +19,7 @@ void vfsHDDManager::CreateEntry(vfsHDD_Entry& entry)
 	CreateBlock(entry);
 }
 
```
Can we have an enum with error code and use it? Would make code easier to read.

tambry replied:
@vlj Sounds like it would make sense. I'll also make some return codes for the virtual memory helper functions.

vlj commented on rpcs3/Emu/HDD/HDD.cpp:
```diff
@@ -99,50 +107,99 @@ void vfsHDDFile::RemoveBlocks(u64 start_block)
 	}
 }
 
```
Same

Ekaseo commented:
im not sure, but it looks like this PR will break the arkedo series games (they will reach the menu, but wont go ingame)

vlj commented:
Why?

danilaml commented:
@Ekaseo is there any new errors in the log? maybe before it was failing silently for some reason.

vlj commented:
Is there any progress on this ?

tambry commented:
@vlj It's rebased an ready for merge, from my point of view.

danilaml commented:
Is it ready for merging, @Nekotekina , or there is still something left?

Ekaseo commented:
its been a while since i tested some builds, but unfortunately, nothing except a few samples work right now... do you ppl even test the stuff before pull it? (its not this PR that broke almost everything, it was something pulled before...) for me, only cube samples work right now, everything else has an acess violation. If its only a problem with my pc, i will include some log files.

tambry commented:
@Ekaseo Same situation here. Almost everything dies with an access violation, except some of the most simple samples.

Nekotekina commented:
What PR caused it?

tambry commented:
@Nekotekina #1274 seems to have broken almost everything.

vlj commented:
Which backend? 
I often test samples like vertex attributes, dice and primitive to test my dx12 pr,  I didn't have access violation with them.

danilaml commented:
@Ekaseo some men just want to watch the world burn :fire: : https://github.com/RPCS3/rpcs3/pull/1245
:smile:

danilaml commented:
@Ekaseo Also, don't forget to provide your settings. For example dx12 crashes if write color/depth buffer is enabled while oGL needs it to display many things correctly.

tambry commented:
@vlj
After merge of #1274, with the following settings:
http://puu.sh/l4qWH/96a32ba0aa.png and http://puu.sh/l4qZo/aecbecc249.png
ARKEDO SERIES - 02 SWAP! (NPEB00614) doesn't work

But on #1272, with the **EXACTLY** same settings:
http://puu.sh/l4reZ/cfe97e57e1.png and http://puu.sh/l4rfd/208ba8df97.png
ARKEDO SERIES - 02 SWAP! (NPEB00614) works

Ekaseo commented:
i use opengl and i use basic settings. i dont have anything special enabled, not even color/depth buffers. i mostly use the lowest resolution. i have vsynch one + framelimiter on auto. null on audio, have camera disabled, im using the default folder paths. I guess opengl has no support now compared to dx12.

tambry commented:
@Ekaseo Well, OpenGL does seem to be a bit of a mess. But I encounter the same errors with DX12.

vlj commented:
I'm intending to factor code between dx12 and GL very soon so that GL backend benefits from the work puts in dx12.

vlj commented:
tambry, I use default configuration in miscalleous parameters.
Per game config is completly broken now.

tambry commented:
@vlj I wasn't even aware of per game config now existing. Looks cool though. According to settings though, I do use the default configuration. (http://puu.sh/l4vDi/548f759efc.png)

vlj commented:
Still no merge ?

Nekotekina commented:
It's conceptually bloating and introduces some errors.

Most of those functions results are assumed to be "OK" and introducing so much code to "report" an error and increase complexity returning... I don't like it. I would use CHECK_ASSERTION macro for them (it's unrelated to `assert` and always compiled).

Nekotekina commented on Utilities/VirtualMemory.cpp:
```diff
@@ -15,27 +16,30 @@ namespace memory_helper
 #ifdef _WIN32
-		return VirtualAlloc(NULL, size, MEM_RESERVE, PAGE_NOACCESS);
+		void* ret = VirtualAlloc(NULL, size, MEM_RESERVE, PAGE_NOACCESS);
+		CHECK_ASSERTION(ret == NULL);
```
CHECK_ASSERTION accepts "right" condition, in your code it's inverted everywhere.
Plus many return conditions are actually wrong.

tambry commented:
@Nekotekina Should be fixed now. If there are any other issues, please let me know.

Nekotekina commented:
You compare unsigned values which cannot be < 0

tambry commented:
@Nekotekina Seems like most of those comparisons were introduced by function return type being unsigned, while the return values were negative. I have fixed those return types in a few places and also added the CHECK_ASSERTION in a few more places.

Nekotekina commented:
Who asked you to change return type?
http://stackoverflow.com/questions/2711522/what-happens-if-i-assign-a-negative-value-to-an-unsigned-variable

```
CHECK_ASSERTION(x.seek(...) != -1);
```

Much better and always works.

Nekotekina commented on rpcs3/Crypto/unself.cpp:
```diff
@@ -890,8 +899,9 @@ bool SELFDecrypter::DecryptNPDRM(u8 *metadata, u32 metadata_size)
-		memcpy(npdrm_key, klicensee_key, 0x10);
+	// FIXME: Check is always false.
+	/*if (klicensee_key != NULL)
+		memcpy(npdrm_key, klicensee_key, 0x10);*/
```
Just don't touch it.

tambry replied:
@Nekotekina What do you mean? This check is **always** false, thus I commented it and added a "FIXME" note. If you want me to uncomment me it, I can do that.

danilaml replied:
It better be kept commented (though it could use `//` comments) with remark. If we keep it the old way everyone will forget about what this check intended to do.
Maybe more elaborate comment would be better.

tambry replied:
@danilaml Should be better now.

danilaml replied:
@tambry how?

Nekotekina commented on Utilities/File.cpp:
```diff
@@ -327,7 +327,13 @@ int OSCopyFile(const char* source, const char* destination, bool overwrite)
+	{
+		return -1;
+	}
+
```
This one is fixed correctly on my branch.

Nekotekina commented on rpcs3/Emu/HDD/HDD.cpp:
```diff
@@ -392,23 +419,28 @@ bool vfsHDD::SearchEntry(const std::string& name, u64& entry_block, u64* parent_
+	}
 
 	m_cur_dir_block = entry.data_block;
 	ReadEntry(m_cur_dir_block, m_cur_dir);
```
No return value

tambry replied:
@Nekotekina Fixed.

Nekotekina commented on Utilities/rPlatform.cpp:
```diff
@@ -54,7 +56,16 @@ std::string rPlatform::getConfigDir()
 		dir = dir + "/rpcs3/";
-		mkdir(dir.c_str(), 0777);
+
+		s32 ret = mkdir(dir.c_str(), 0777);
```
It may return -1, not an error code. Also use `std::printf` to report success/failure, LOG_ won't work correctly.

Nekotekina commented on rpcs3/Emu/SysCalls/Modules/cellResc.cpp:
```diff
@@ -344,14 +344,17 @@ s32 CalculateSurfaceByteSize(u32 mode, CellRescDsts *dsts)
+			if ((index = GetRescDestsIndex(bufMode)) != 0)
+			{
+				oneBufSize = CalculateSurfaceByteSize(bufMode, &(s_rescInternalInstance->m_rescDsts[index]));
+			}
```
1) There shouldn't be any defect there, from my understanding.
2) Fix doesn't make sense

Nekotekina commented on rpcs3/Emu/Memory/MemoryBlock.h:
```diff
@@ -82,7 +84,13 @@ class VirtualMemoryBlock
+
+		if (!getRealAddr(addr, realAddr))
+		{
+			LOG_ERROR(HLE, "Getting the real address failed. (addr=%u)", addr);
```
0x%x

tambry commented:
@Nekotekina Are there any other issues blocking this from being merged?
